doc/fileformat/serdes.rst

```diff
@@ -0,0 +1,262 @@
+    // [T, 2T) should be identical.
+
+Note that we left the definition of ``io`` open, as ``serializer`` uses it
+through a well-defined interface (see next section). Thus, one can simply add
+new implementations. Arbor currently ships with ``arborio::json_serdes`` that
+produces JSON output, more might be added over time.
+
+In Python, currently only support for (de)serialization from/to JSON strings is
+offered.
+
+  .. code:: python
+
+    import arbor as A
+
+    rec = my_recipe()
+    sim = A.simulation(rec)
+    jsn = sim.serialize()
+    sim.deserialize(jsn)
+
+
+Writing your own Storage Engine (C++ only)
+------------------------------------------
+
+A storage engine is used to construct the serializer and is responsible for
+writing out the data to whatever format and location required. Currently Arbor
+offers a JSON engine in ``arborio/json_serdes.hpp`` which produces a JSON value
+in memory. The serializer is polymorphic in the actual engine, which is only
+require to implement the following interface.
+
+   .. code:: c++
+
+         struct interface {
+             virtual void write(const key_type&, std::string) = 0;
+             virtual void write(const key_type&, double) = 0;
+             virtual void write(const key_type&, long long) = 0;
+             virtual void write(const key_type&, unsigned long long) = 0;
+
+             virtual void read(const key_type&, std::string&) = 0;
+             virtual void read(const key_type&, double&) = 0;
+             virtual void read(const key_type&, long long&) = 0;
+             virtual void read(const key_type&, unsigned long long&) = 0;
+
+             virtual std::optional<key_type> next_key() = 0;
+
+             virtual void begin_write_map(const key_type&) = 0;
+             virtual void end_write_map() = 0;
+             virtual void begin_write_array(const key_type&) = 0;
+             virtual void end_write_array() = 0;
+
+             virtual void begin_read_map(const key_type&) = 0;
+             virtual void end_read_map() = 0;
+             virtual void begin_read_array(const key_type&) = 0;
+             virtual void end_read_array() = 0;
+
+             virtual ~interface() = default;
+         };
+
+The ``read`` and ``write`` methods are responsible for inserting and extracting
+the relevant items. The ``begin_write_array`` and ``end_write_array`` methods
+bracket a write of an array value and announce that the following keyes are to
+be interpreted as integer indices. Analogous for the ``map`` counterparts and
+the associated ``begin_read`` and ``end_read`` methods. Finally, ``next_key`` is
+used during reading of containers to retrieve an optional next key and advanced
+the internal iterator. If empty, the container is exhausted, else the contained
+key can be used to retrieve the associated value. See examples below and the JSON
+interface in ``arborio``.
+
+
+Adding Snapshotting to new Objects (C++ only)
```

Comment:
Swap this section with `Writing your own Storage Engine`. This is more relevant to most users.

Reply:
Why couldn't a user add new serializable objects to a simulation?

Reply:
To the `simulation` _type_ inside libarbor?

Reply:
Serializing a simulation object is where things begin, right? I assume any object in it, will be serialized if they have implemented `read`/`write` or use of of the `ARB_SERDES_*` macros. Should somebody add a new type like that, and have a recipe using the type, then it'll be included in the serialization of the simulation object, right?

Reply:
No, `recipe` is by definition not serializable. We can -- in general -- not serialize a function except dumping its source code / assembly, which we cannot access from inside C++ (not practically, that is). In that sense `recipe`
_is_ its own serialization.

Thus the comment in the docs about 'treating recipe + main + serialize as one bundle'. This and only this is a 
complete representation.

A user cannot add an arbitrary object to a simulation at C++ level. This checkpoint relies on us to identify and
annotate the parts to serialize. See here on what we would need for making such functionality:
https://herbsutter.com/2017/07/26/metaclasses-thoughts-on-generative-c/

Python is a different story, but we'll not scan our object tree to find user data and speculatively dump it for them.

Reply:
Right. Discussed over VC.